### DISCUSSION TITLE
Conditional impersonation

### DISCUSSION THREAD

**Discussion by SebastianSchenk** [keycloak/keycloak#20252] [Q&A]
My goal is to allow certain users (support) to impersonate a certain type of other users (customers). I tried the /impersonate endpoint, as well as the token exchange feature. 

However, both options require the user to have permission to impersonate. Either via role or fine grain admin permissions. 
This means the user could impersonate any other user in the realm. Even users with higher permissions like an admin.

I know that I can specify which users are allowed to be impersonated with the `user-impersonated` permission. Unfortunately, this is then applied to all users that can impersonate. Event the admin users that should be allowed to impersonate everyone.
I wrote my own js-policy for the `user-impersonated` permission, but in this context I have no access to the user who is initiating the request. Only to the user that is being impersonated. So this doesn't help much. In general, a policy executed in the context of `user-impersonated` gets way less information passed down than other permissions. I don't even have access to the client id, which is normally accessible via `kc.client.id` ([documented here](https://www.keycloak.org/docs/latest/authorization_services/index.html#the-evaluation-context)).

Using the token exchange feature I have the option to omit the `subject_token` parameter and therefor trigger the impersonation in the name of the client. All I have to do to make this work is to define a policy that allows the client to impersonate users. This does work, but in this scenario I lose the information in the token, which user initially triggered the impersonation. Normally this information is being added to the token if the corresponding scope mappers are enabled. Besides that, this method is super risky according to the documentation [here](https://www.keycloak.org/docs/latest/securing_apps/#direct-naked-impersonation).

So even though Keycloak provides different methods to restrict impersonation, in the end it is a all or nothing situation in my opinion. The [documentation](https://www.keycloak.org/docs/latest/securing_apps/#impersonation) says "For internal and external token exchanges, the client can request on behalf of a user to impersonate a different user. For example, you may have an admin application that needs to impersonate a user so that a support engineer can debug a problem." On one hand, if I give support engineers the permission to impersonate they can impersonate any user, not only the customers. On the other hand, if I restrict users that can be impersonated I also limit any other admin user.

Has anyone an idea how I can archive my goal with the currently implemented methods in Keycloak?

**Comment by flyinfish** [comment]
i came upon this when trying to explore "impersonation". 
it looks to me as it is a feature missing.

i illustrated an example at https://github.com/flyinfish/keycloak-examples/blob/main/impersonation/README.md with the issue above at [#s3---conditional-impersonation](https://github.com/flyinfish/keycloak-examples/blob/main/impersonation/README.md#s3---conditional-impersonation).

@SebastianSchenk did you come to a solution?
@gdelbos as stated in https://github.com/keycloak/keycloak/discussions/26502#discussioncomment-9678116 you solved something similar with a custom token-exchange-provider as i understand. looks like you could solve it out-of-the-box when we had the missing feature described here 
@patriot1burke do you like to state your opinion?

thx to all

**Reply by SebastianSchenk** [reply]
@flyinfish unfortunately, no. But I'm still interested to find a solution.

**Reply by gdelbos** [reply]
Hi @SebastianSchenk 

You may have a look at my comment linked by @flyinfish 

Regarding the token exchange provider, I've just gone with extending the default one (`org.keycloak.protocol.oidc.DefaultTokenExchangeProvider`), overriding only the `tokenExchange` method by doing the following change.

```diff
- if (!AdminPermissions.evaluator(session, realm, auth).users().canImpersonate(requestedUser, client)) {
+ var user = AdminPermissions.evaluator(session, realm, auth).users();
+ if (!user.canImpersonate(requestedUser, client) || !user.canView(requestedUser)) {
```

This change is only used in case a user is doing an impersonation.
Using this modification and some permission on user groups, you will be able to give, for instance, 
- Alice the authorization to impersonate Jane but not John
- Rebecca the authorization to impersonate John but not Jane.

I can give you more detail on the setup if you wish.

**Reply by flyinfish** [reply]
thanks @gdelbos good input. 
what i like to do is tackle requirement as yours out-of-the box without extending

**Reply by flyinfish** [reply]
since no more input i filed
https://github.com/keycloak/keycloak/issues/31557
